File: src/main/java/no/nav/bidrag/beregn/forskudd/rest/exception/ExceptionHandlerAdvice.java
package no.nav.bidrag.beregn.forskudd.rest.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class ExceptionHandlerAdvice {

  @ExceptionHandler
  public ResponseEntity<?> handleUgyldigInputException(UgyldigInputException exception) {
    return ResponseEntity
        .status(HttpStatus.BAD_REQUEST)
        .header("Error", errorMsg(exception))
        .build();
  }

  @ExceptionHandler
  public ResponseEntity<?> handleSjablonConsumerException(SjablonConsumerException exception) {
    return ResponseEntity
        .status(exception.getStatusCode())
        .header("Error", errorMsg(exception))
        .build();
  }

  private String errorMsg(RuntimeException runtimeException) {
    return String.format("%s: %s", runtimeException.getClass().getSimpleName(), runtimeException.getMessage());
  }
}

File: src/main/kotlin/no/nav/bidrag/beregn/forskudd/rest/exception/ExceptionHandlerAdvice.kt
package no.nav.bidrag.beregn.forskudd.rest.exception

import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.client.HttpStatusCodeException

@ControllerAdvice
class ExceptionHandlerAdvice {
    @ExceptionHandler
    fun handleUgyldigInputException(exception: IllegalArgumentException): ResponseEntity<*> {
        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .header(HttpHeaders.WARNING, errorMsg(exception))
            .build<Any>()
    }

    @ExceptionHandler
    fun handleSjablonConsumerException(exception: HttpStatusCodeException): ResponseEntity<*> {
        return ResponseEntity
            .status(exception.statusCode)
            .header(HttpHeaders.WARNING, errorMsg(exception))
            .build<Any>()
    }

    private fun errorMsg(runtimeException: RuntimeException): String {
        return String.format("%s: %s", runtimeException.javaClass.simpleName, runtimeException.message)
    }
}
